api/src/main/java/org/eclipse/microprofile/config/spi/ConfigProviderResolver.java

```diff
@@ -105,39 +103,25 @@ public static ConfigProviderResolver instance() {
                 if (instance != null) {
                     return instance;
                 }
+                // first try my class loader
+                ConfigProviderResolver instance = loadSpi(ConfigProviderResolver.class.getClassLoader(), null);
+                // next try the system class loader
+                instance = loadSpi(null, instance);
```

Comment:
Your change will bypass the ThreadContextClassloader, as you try to use the classloader from ConfigProviderResolver or the system classloader, which might not the right classloader.

Reply:
@struberg @OndroMih Can you please check against your implementation?

Reply:
That is correct. There *is* no correct thread context classloader in a static initializer, because there is no way to guarantee who the caller is - it can vary between runs even.  It is never correct to reference TCCL or any thread local value from a static initializer.

Reply:
It makes sense, @dmlloyd. When I look at it now, @Emily-Jiang, I think it's enough to have a single ConfigProviderResolver instance per JVM as it doesn't hold any state. In the end, the concrete implementation of `ConfigProviderResolver` can load the Config instance using the thread context classloader as it's done in Payara already: https://github.com/payara/Payara/blob/master/nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/spi/ConfigProviderResolverImpl.java. I would have to review it more thoroughly but I think that Payara doesn't expect that the implementation of ConfigProviderResolver is loaded by any specific classloader.

Reply:
We've tested this PR with the current Payara implementation and all seems to work. It passes the MP TCK. So we approve this PR.